distributed/shuffle/multi_file.py

```diff
@@ -0,0 +1,286 @@
+        The maximum size of a single message that we want to send
+
+    Parameters
+    ----------
+    directory: pathlib.Path
+        Where to write and read data.  Ideally points to fast disk.
+    dump: callable
+        Writes an object to a file, like pickle.dump
+    load: callable
+        Reads an object from that file, like pickle.load
+    join: callable
+        Joins many objects together
+    send: callable
+        How to send a list of shards to a worker
+    sizeof: callable
+        Measures the size of an object in memory
+    """
+
+    memory_limit = parse_bytes("1 GiB")
+    _queues: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+    concurrent_files = 2
+    total_size = 0
+
+    def __init__(
+        self,
+        directory,
+        dump=pickle.dump,
+        load=pickle.load,
+        join=None,
+        sizeof=sizeof,
+        loop=None,
+    ):
+        assert join
+        self.directory = pathlib.Path(directory)
+        if not os.path.exists(self.directory):
+            os.mkdir(self.directory)
+        self.dump = dump
+        self.load = load
+        self.join = join
+        self.sizeof = sizeof
+
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self.total_size = 0
+        self.total_received = 0
+
+        self.condition = asyncio.Condition()
+
+        self.bytes_written = 0
+        self.bytes_read = 0
+
+        self._done = False
+        self._futures = set()
+        self.active = set()
+        self.diagnostics = defaultdict(float)
+
+        self._communicate_future = asyncio.ensure_future(self.communicate())
+        self._loop = loop
+
+    @property
+    def queue(self):
+        try:
+            return MultiFile._queues[self._loop]
+        except KeyError:
+            queue = asyncio.Queue()
+            for _ in range(MultiFile.concurrent_files):
+                queue.put_nowait(None)
+            MultiFile._queues[self._loop] = queue
+            return queue
+
+    async def put(self, data: dict[str, list[object]]):
+        """
+        Writes many objects into the local buffers, blocks until ready for more
+
+        Parameters
+        ----------
+        data: dict
+            A dictionary mapping destinations to lists of objects that should
+            be written to that destination
+        """
+        this_size = 0
+        for id, shard in data.items():
+            size = self.sizeof(shard)
+            self.shards[id].extend(shard)
+            self.sizes[id] += size
+            self.total_size += size
+            MultiFile.total_size += size
+            self.total_received += size
+            this_size += size
+
+        del data, shard
+
+        while MultiFile.total_size > self.memory_limit:
+            with self.time("waiting-on-memory"):
+                async with self.condition:
+
+                    try:
+                        await asyncio.wait_for(
+                            self.condition.wait(), 1
```

Comment:
I don't like this hard coded timeout. This should not silently time out. The very least would be a warning log. I think I'd prefer not having a timeout since if this does not clear, the system is somehow broken.

Reply:
It's in a whiile loop.  I just want to check in with total_size from time to to time to see if something else has happened.

Reply:
Also, let's move this back to the previous PR if possible.